functional/lenses.md

```diff
@@ -0,0 +1,481 @@
+lenses.
+
+While pure functional languages have many more uses, a primary use for this in
+Rust is to split problems of recursion and types into pieces that can be
+composed.
+A good example of this is the traits in Serde.
+
+Trying to understand the way Serde works by reading just the API is quite a
+challenge.
+Consider the `Deserializer` trait:
+
+```rust,ignore
+pub trait Deserializer<'de>: Sized {
+    type Error: Error;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>;
+
+    fn deserialize_bool<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>;
+
+    // remainder ommitted
+}
+```
+
+Wow, that's a lot of generics on functions, in addition to a type on the trait
+itself!
+Why are they all there?
+
+They are there because this API is a prism: an attempt to make a generic set of
+"lenses" work across a generic set of types for "observation."
+
+To understand it better, consider what each "lense" looks like.
+Here is the definition of the `Visitor` type:
+
+```rust,ignore
+pub trait Visitor<'de>: Sized {
+    type Value;
+
+    fn expecting(&self, formatter: &mut Formatter<'_>) -> Result;
+
+    fn visit_bool<E>(self, v: bool) -> Result<Self::Value, E>
+    where
+        E: Error;
+
+    // remainder omitted
+}
+```
+
+This is similar to `BalanceChange` earlier. It is acting as a lens over a
+generic and unknown type of data.
+Its operation is to support composing a set of uniform data structures from
+that data represented by Serde's `Value` type.
+
+But unlike `BalanceChange` it's a *trait*.
+It's a *family* of lenses.
+
+For example, consider the identity record from earlier but simplified:
+
+```json
+{ "name": "Jane Doe",
+  "customer_id": 1048576332,
+}
+```
+
+How would the Serde library transform this JSON into `struct CreditRecord`?
+
+It would create a `Visitor` "lens", which would then try to "observe" expected
+data through a series of calls:
+
+1. Visit the start of a map (the `{}` in Serde terminology).
+1. Visit a key string called "name".
+1. Visit an unknown string which is the value of the `name` field.
+1. Visit a key string called "customer_id".
+1. Visit an unknown string which is the value of the `customer_id` field.
+1. Expect the end of the map.
+1. Visit complete.
+
+If the wrong "observation" happens, an error is thrown and deserialization
+fails.
+
+With each new type, a new implementation of the `Visitor` trait is needed,
+a type to explain what to observe in what order to fill out the struct.
+While Rust does not support generic traits as noted earlier, Serde solves this
+usability problem with a derive procedural macro:
+
+```rust,ignore
+use serde::Deserialize;
+
+#[derive(Deserialize)]
+struct IdRecord {
+    name: String,
+    customer_id: String,
+}
+```
+
+In C++, the equivalent would be to generate a very complex code template, with
+a lot of compile time logic -- if it's even possible.
```

Comment:
Yes, it is ;) But it may become virtually unreadable at some point and at least in the past there was a non-vanishing probability that you stumbled over differences between implementations (e.g. clang vs. gcc).

Reply:
True. In C++, anything is possible :smile: